File: red/test/storage.test.ts
import {afterEach, expect, spyOn, test} from "bun:test";
import * as tofu from "red/tofu";
import {runtime} from "red/runtime";
import * as storage from "../src/storage.ts";
let mocked: ReturnType<typeof spyOn> | undefined;
afterEach(() => mocked?.mockRestore());
const opts = {profile:"owned",workdir:"/tmp", "langfuse-storage-managed":true,"neon-r2-bucket":"owned-neon", "langfuse-s3-bucket":"owned-data", "langfuse-backup-r2-bucket":"owned-backup", "neon-r2-region":"us-east-1"};
test("managed storage refuses existing or inaccessible untracked buckets", async () => {
  for (const probe of [{exit:0,out:"",err:""},{exit:1,out:"",err:"(403) Forbidden"}]) {
    mocked = spyOn(runtime,"exec").mockResolvedValueOnce({exit:0,out:"",err:""}).mockResolvedValueOnce({exit:0,out:"",err:""}).mockResolvedValue(probe);
    await expect(storage.ownershipPreflight(opts)).rejects.toThrow("refuses to adopt");
    mocked.mockRestore();
  }
});
test("fresh remote state probes three application buckets", async () => {
  mocked = spyOn(runtime,"exec").mockResolvedValueOnce({exit:0,out:"",err:""}).mockResolvedValueOnce({exit:1,out:"",err:"No state file was found!"}).mockResolvedValue({exit:1,out:"",err:"(404) Not Found"});
  await storage.ownershipPreflight(opts);
  expect(mocked).toHaveBeenCalledTimes(5);
});
test("renaming a tracked bucket refuses adoption of an existing destination", async () => {
  mocked = spyOn(runtime,"exec").mockResolvedValueOnce({exit:0,out:"",err:""}).mockResolvedValueOnce({exit:0,out:'aws_s3_bucket.application["neon"]',err:""}).mockResolvedValueOnce({exit:0,out:JSON.stringify({values:{root_module:{resources:[{address:'aws_s3_bucket.application["neon"]',values:{bucket:"old-neon"}}]}}}),err:""}).mockResolvedValue({exit:0,out:"",err:""});
  await expect(storage.ownershipPreflight(opts)).rejects.toThrow("refuses to adopt");
});
test("each role receives a distinct scoped environment pair", () => {
  const credentials = {credentials:Object.fromEntries(["neon","data","backup"].map(role=>[role,{access_key_id:role+"-id",secret_access_key:role+"-secret"}]))};
  const env = storage.credentialEnv({...opts,"langfuse/storage-credentials":credentials});
  expect(env.COLORS_PAR_NEON_R2_ACCESS_KEY_ID).toBe("neon-id");
  expect(env.COLORS_PAR_LANGFUSE_STORAGE_R2_ACCESS_KEY_ID).toBe("data-id");
  expect(env.COLORS_PAR_LANGFUSE_BACKUP_R2_ACCESS_KEY_ID).toBe("backup-id");
  expect(env.AWS_ACCESS_KEY_ID).toBeUndefined();
  expect(()=>storage.credentialEnv(opts)).toThrow("unavailable");
});

test("sensitive tofu JSON decodes into three scoped Ansible credential pairs", async () => {
  const value = Object.fromEntries(["neon","data","backup"].map(role => [role,{access_key_id:role+"-id",secret_access_key:role+"-secret"}]));
  mocked = spyOn(runtime,"exec").mockResolvedValue({exit:0,out:JSON.stringify({credentials:{sensitive:true,type:["object",{}],value}}),err:""});
  const decoded = await tofu.outputs("/unused");
  const env = storage.credentialEnv({...opts,"langfuse/storage-credentials":decoded});
  expect(env.COLORS_PAR_NEON_R2_ACCESS_KEY_ID).toBe("neon-id");
  expect(env.COLORS_PAR_LANGFUSE_STORAGE_R2_SECRET_ACCESS_KEY).toBe("data-secret");
  expect(env.COLORS_PAR_LANGFUSE_BACKUP_R2_SECRET_ACCESS_KEY).toBe("backup-secret");
  expect(mocked.mock.calls[0]![0]).toEqual(["tofu","output","-json"]);
});

File: green/test/clj/io/github/getcolors/langfuse/storage_test.clj
(ns io.github.getcolors.langfuse.storage-test
  (:require [clojure.test :refer [deftest is]]
            [clojure.java.shell :as shell]
            [green.tofu :as tofu]
            [cheshire.core :as json]
            [green.process :as process]
            [io.github.getcolors.langfuse.storage :as storage]
            [io.github.getcolors.langfuse.validate :as validate]
            [io.github.getcolors.langfuse.validate-test :refer [base]]))
(def managed (assoc base :provider-backend "s3" :s3-bucket "langfuse-state" :s3-region "us-east-1"
                        :langfuse-storage-managed true :langfuse-storage-provider "s3"
                        :neon-r2-bucket "langfuse-neon" :neon-r2-region "us-east-1"
                        :langfuse-backup-r2-region "us-east-1"))
(def credentials {:credentials (into {} (for [role [:neon :data :backup]]
                                         [role {:access_key_id (str (name role) "-id") :secret_access_key (str (name role) "-secret") }]))})
(deftest managed-s3-validation-and-credential-requirements
  (is (empty? (validate/state-errors (dissoc managed :r2-bucket :r2-endpoint))))
  (doseq [change [{:langfuse-storage-provider "r2"} {:langfuse-storage-managed "true"}
                  {:neon-r2-region "auto"} {:langfuse-backup-r2-region "eu-west-1"}
                  {:neon-r2-bucket (:langfuse-s3-bucket managed)}
                  {:s3-bucket (:langfuse-s3-bucket managed)}]]
    (is (seq (validate/state-errors (merge managed change)))))
  (is (= #{"required credential is not set: COLORS_PAR_CLOUDFLARE_API_TOKEN"
           "required credential is not set: COLORS_PAR_LANGFUSE_ENCRYPTION_KEY"
           "required credential is not set: COLORS_PAR_LANGFUSE_SALT"
           "required credential is not set: COLORS_PAR_LANGFUSE_INIT_USER_PASSWORD"}
         (set (validate/secret-errors managed :create)))))
(deftest three-bucket-identities-reach-the-existing-secret-lookups
  (let [environment (storage/credential-env (assoc managed :langfuse/storage-credentials credentials))]
    (is (= "neon-id" (get environment "COLORS_PAR_NEON_R2_ACCESS_KEY_ID")))
    (is (= "data-secret" (get environment "COLORS_PAR_LANGFUSE_STORAGE_R2_SECRET_ACCESS_KEY")))
    (is (= "backup-id" (get environment "COLORS_PAR_LANGFUSE_BACKUP_R2_ACCESS_KEY_ID")))
    (is (nil? (get environment "AWS_ACCESS_KEY_ID"))))
  (is (thrown? Exception (storage/credential-env managed))))
(deftest ownership-rejects-untracked-existing-buckets-and-unknown-state
  (doseq [probe [{:exit 0 :out ""} {:exit 1 :err "(403) Forbidden"}]]
    (with-redefs [process/run (fn [args _] (if (= "aws" (first args)) probe {:exit 0 :out ""}))]
      (is (thrown? Exception (storage/ownership-preflight! managed)))))
  (with-redefs [process/run (fn [args _] (if (= args ["tofu" "state" "list"]) {:exit 1 :err "AccessDenied"} {:exit 0 :out ""}))]
    (is (thrown? Exception (storage/ownership-preflight! managed)))))
(deftest first-create-probes-all-three-buckets
  (let [probes (atom [])]
    (with-redefs [process/run (fn [args _]
                               (cond
                                 (= args ["tofu" "state" "list"]) {:exit 1 :err "No state file was found!"}
                                 (= "aws" (first args)) (do (swap! probes conj (nth args 4)) {:exit 254 :err "(404) Not Found"})
                                 :else {:exit 0 :out ""}))]
      (storage/ownership-preflight! managed)
      (is (= #{"langfuse-neon" "langfuse-storage" "langfuse-backup"} (set @probes))))))
(deftest tracked-address-must-also-match-bucket-name
  (with-redefs [process/run (fn [args _]
                             (cond
                               (= args ["tofu" "state" "list"]) {:exit 0 :out "aws_s3_bucket.application[\"neon\"]"}
                               (= args ["tofu" "show" "-json"]) {:exit 0 :out (json/generate-string {:values {:root_module {:resources [{:address "aws_s3_bucket.application[\"neon\"]" :values {:bucket "foreign-old-name"}}]}}})}
                               :else {:exit 0 :out ""}))]
    (is (thrown? Exception (storage/ownership-preflight! managed)))))
(deftest adopted-storage-does-not-provision
  (with-redefs [process/run (fn [& _] (throw (AssertionError. "adopted storage must not provision")))]
    (is (= 0 (:green/exit (storage/step base))))))

(deftest sensitive-tofu-json-reaches-ansible-with-nested-string-keys
  ;; Exercise the real SDK decoder: only the top-level output key is keywordized.
  (let [wire (json/generate-string
               {:credentials {:sensitive true :type ["object" {}] :value (:credentials credentials)}})
        calls (atom [])]
    (with-redefs [shell/sh (fn [& args]
                            (swap! calls conj (take 3 args))
                            {:exit 0 :out wire :err ""})]
      (let [decoded (tofu/outputs "/unused")
            env (storage/credential-env (assoc managed :langfuse/storage-credentials decoded))]
        (is (contains? (:credentials decoded) "neon"))
        (is (= "neon-id" (get env "COLORS_PAR_NEON_R2_ACCESS_KEY_ID")))
        (is (= "data-secret" (get env "COLORS_PAR_LANGFUSE_STORAGE_R2_SECRET_ACCESS_KEY")))
        (is (= "backup-secret" (get env "COLORS_PAR_LANGFUSE_BACKUP_R2_SECRET_ACCESS_KEY")))
        (is (= [["tofu" "output" "-json"]] @calls))))))
